lib/stages/BeamAssemble.cpp

```diff
@@ -0,0 +1,207 @@
+                            in_frame[0] & 0x0F, (in_frame[0] & 0xF0) >> 4 );
+                }
+            }
+        }
+        else {
+            // increment the number of accepted frame from all incoming beam data frames
+            accepted_beam_frames_count++;
+
+            // determine which output frame to write on based on the timestamp: if this is a new time
+            // stamp, make a new entry in the map and acquire an empty frame from the output buffer
+            if (out_frame_map.find(current_frame_timestamp) == out_frame_map.end()) {
+                // wait for the first available empty frame
+                out_frame = wait_for_empty_frame(out_buf, unique_name.c_str(), out_frame_id);
+                if (out_frame == nullptr) {
+                    // no output frame buffer available, so leave the thread infinite loop
+                    break;
+                }
+
+                // allocate a new metadata object for this empty output buffer frame
+                allocate_new_metadata_object(out_buf, out_frame_id);
+
+                // copy the base metadata to this frame metadata
+                copy_base_to_frequency_assembled_Metadata( metadata, get_metadata(out_buf, out_frame_id));
+
+                // add the new output frame to the map
+                out_frame_map[current_frame_timestamp] = std::make_pair( out_frame, out_frame_id);
+                out_frame_id++;
+
+                // zero all the frame: zeros implies missing data as well
+                memset(out_frame, 0, out_buf->frame_size);
+            } else {
+                // retrieve the output frame from the map
+                out_frame = out_frame_map[current_frame_timestamp].first;
+            }
+
+            // loop over received frequencies in the input buffer frame and
+            // put their data in the corresponding output buffer frame
+            for (uint32_t f = 0; f < num_freq_per_stream; ++f) {
+                // get the frequency id
+                auto freq_id = Telescope::instance().to_freq_id(metadata->stream_id, f);
+
+                // input frame buffer frequency data size
+                auto in_frame_freq_size = in_buf->frame_size / num_freq_per_stream;
+
+                // determine the offset based on the frequency id
+                auto out_frame_offset = freq_id * (out_buf->frame_size - num_freq_per_output_frame) / num_freq_per_output_frame;
+                // is that the same as: freq_id * in_frame_freq_size ?
+                // assert(out_frame_offset == freq_id * in_frame_freq_size);
+
+                // copy the data from in_frame to the out_frame + some offset with no change
+                // Note that the first num_freq_per_output_frame is the frequency id indicator,
+                // which is 0 or 1 indicating the frequency data copied to the output
+                memcpy(out_frame + num_freq_per_output_frame + out_frame_offset, in_frame, in_frame_freq_size);
```

Comment:
For `num_freq_per_stream > 1` the in_frame layout does not follow this pattern.  The pattern for multi-frequency beams is (time, frequency, pol), where each byte is still a complex 4+4-bit [Re,Im] offset encoded pair.  And `size(pol) = 2`, `size(frequency) = num_freq_per_stream`, and `size(time) = samples_per_data_set`.

Reply:
To be clear, `(time, frequency, pol)` is slowest to fastest varying index.

Reply:
Thanks, I fixed it.